### PR TITLE
Propagate serialization options in diagnostics

### PR DESCRIPTION
**Pull request by rcj-siteimprove**
Propagate the serialization options in `Diagnostic` subclasses so that they are eventually applied when serializing `Element`s.

I have deliberately only propagated the options where they would eventually be passed to `Element.toJSON`. Propagating them to all sub-calls would explode and the options wouldn't be used for anything anyway, so it seems somewhat a waste of time.

### CODE REVIEW

**Comment by rcj-siteimprove** [comment]
!pr extract

**Comment by Jym77** [comment]
Let's go a bit back on the changes.
Instead of adding the type parameter everywhere, we can keep it on the `toJSON()` method. Also need to keep sure all properties are optionals.

```typescript
class Diagnostic {
  …
  public toJSON(options: json.SerializationOptions) { … }
}

class DiagWithNodeAndFooAndBar {
  …
  public toJSON(options: Node.SerializationOptions & Foo.SerializationOptions & Bar.SerializationOptions) { … }
```

This does, however require we also change the existing ones for consistency.

**Comment by rcj-siteimprove** [comment]
!pr extract

**Comment by rcj-siteimprove** [comment]
> Let's go a bit back on the changes. Instead of adding the type parameter everywhere, we can keep it on the `toJSON()` method. Also need to keep sure all properties are optionals.
> 
> ```ts
> class Diagnostic {
>   …
>   public toJSON(options: json.SerializationOptions) { … }
> }
> 
> class DiagWithNodeAndFooAndBar {
>   …
>   public toJSON(options: Node.SerializationOptions & Foo.SerializationOptions & Bar.SerializationOptions) { … }
> ```
> 
> This does, however require we also change the existing ones for consistency.

@Jym77 This should be done now

File: packages/alfa-collection/src/collection.ts [hunk 1]
```diff
@@ -1,12 +1,8 @@
...
 import type { Array } from "@siteimprove/alfa-array";
 import type { Callback } from "@siteimprove/alfa-callback";
-import type {
-  Comparable,
-  Comparison} from "@siteimprove/alfa-comparable";
-import {
-  type Comparer
-} from "@siteimprove/alfa-comparable";
+import type { Comparable, Comparison } from "@siteimprove/alfa-comparable";
+import { type Comparer } from "@siteimprove/alfa-comparable";
```
Comment (Jym77) [inline packages/alfa-collection/src/collection.ts:5]:
```suggestion
import type { Comparable, Comparer, Comparison } from "@siteimprove/alfa-comparable";
```

File: packages/alfa-list/src/list.ts [hunk 2]
```diff
@@ -1,12 +1,8 @@
...
 import type { Callback } from "@siteimprove/alfa-callback";
 import type { Collection } from "@siteimprove/alfa-collection";
-import type {
-  Comparison} from "@siteimprove/alfa-comparable";
-import {
-  Comparable,
-  type Comparer
-} from "@siteimprove/alfa-comparable";
+import type { Comparison } from "@siteimprove/alfa-comparable";
+import { Comparable, type Comparer } from "@siteimprove/alfa-comparable";
```
Comment (Jym77) [inline packages/alfa-list/src/list.ts:5]:
```suggestion
import { Comparable, type Comparer, type Comparison } from "@siteimprove/alfa-comparable";
```

File: packages/alfa-option/src/option.ts [hunk 3]
```diff
@@ -1,10 +1,6 @@
 import type { Callback } from "@siteimprove/alfa-callback";
-import type {
-  Comparable,
-  Comparison} from "@siteimprove/alfa-comparable";
-import {
-  type Comparer
-} from "@siteimprove/alfa-comparable";
+import type { Comparable, Comparison } from "@siteimprove/alfa-comparable";
+import { type Comparer } from "@siteimprove/alfa-comparable";
```
Comment (Jym77) [inline packages/alfa-option/src/option.ts:3]:
```suggestion
import type { Comparable, Comparison } from "@siteimprove/alfa-comparable";
import { type Comparer } from "@siteimprove/alfa-comparable";
```

File: packages/alfa-sequence/src/cons.ts [hunk 4]
```diff
@@ -1,12 +1,8 @@
...
 import type { Callback } from "@siteimprove/alfa-callback";
 import type { Collection } from "@siteimprove/alfa-collection";
-import type {
-  Comparison} from "@siteimprove/alfa-comparable";
-import {
-  Comparable,
-  type Comparer
-} from "@siteimprove/alfa-comparable";
+import type { Comparison } from "@siteimprove/alfa-comparable";
+import { Comparable, type Comparer } from "@siteimprove/alfa-comparable";
```
Comment (Jym77) [inline packages/alfa-sequence/src/cons.ts:5]:
```suggestion
import { Comparable, type Comparer, type Comparison } from "@siteimprove/alfa-comparable";
```

File: packages/alfa-sequence/src/sequence.ts [hunk 5]
```diff
@@ -1,12 +1,8 @@
...
 import type { Callback } from "@siteimprove/alfa-callback";
 import type { Collection } from "@siteimprove/alfa-collection";
-import type {
-  Comparable,
-  Comparison} from "@siteimprove/alfa-comparable";
-import {
-  type Comparer
-} from "@siteimprove/alfa-comparable";
+import type { Comparable, Comparison } from "@siteimprove/alfa-comparable";
+import { type Comparer } from "@siteimprove/alfa-comparable";
```
Comment (Jym77) [inline packages/alfa-sequence/src/sequence.ts:5]:
```suggestion
import type { Comparable, type Comparer, Comparison } from "@siteimprove/alfa-comparable";
```

File: packages/alfa-slice/src/slice.ts [hunk 6]
```diff
@@ -1,12 +1,8 @@
...
 import type { Callback } from "@siteimprove/alfa-callback";
 import type { Collection } from "@siteimprove/alfa-collection";
-import type {
-  Comparison} from "@siteimprove/alfa-comparable";
-import {
-  Comparable,
-  type Comparer
-} from "@siteimprove/alfa-comparable";
+import type { Comparison } from "@siteimprove/alfa-comparable";
+import { Comparable, type Comparer } from "@siteimprove/alfa-comparable";
```
Comment (Jym77) [inline packages/alfa-slice/src/slice.ts:5]:
```suggestion
import { Comparable, type Comparer, type Comparison } from "@siteimprove/alfa-comparable";
```

**Review by Jym77 — CHANGES_REQUESTED** [review]
Cleaning a few double imports, not sure why tslint messed these up 🤔

**Review by Jym77 — APPROVED** [review]

**Comment by rcj-siteimprove** [comment]
!pr extract

**Review by Jym77 — APPROVED** [review]

**Pull request merged**